Uploader/Program.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using NLog;
+using System.IO;
+using System.Diagnostics;
+using System.Reflection;
+using System.Threading;
+using System.Timers;
+using Newtonsoft.Json;
+using System.IO.Abstractions;
+using System.Collections.Generic;
+
+/// <summary>
+/// Main entry point of the program.
+/// </summary>
+public class Uploader
+{
+    private const long TimerIntervalInMilliseconds = 1000 * 60 * 5;
+    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+
+    private readonly string traceDirectory;
+    private readonly Config config;
+    private readonly TimerAction timerAction;
+
+    /// <summary>
+    /// Main entry point. Expects a single argument: the path to a directory that contains the trace files to upload.
+    /// </summary>
+    public static void Main(string[] args)
+    {
+        if (IsAlreadyRunning())
+        {
+            Console.WriteLine("Another instance is already running.");
+            return;
+        }
+        new Uploader(args).Run();
+    }
+
+    private static bool IsAlreadyRunning()
+    {
+        Process current = Process.GetCurrentProcess();
+        // If this process is not running from the exe file, we allow it
+        // Not sure when this happens, though
+        if (Assembly.GetExecutingAssembly().Location.Replace("/", "\\") != current.MainModule.FileName)
+        {
+            return false;
+        }
+
+        Process[] processes = Process.GetProcessesByName(current.ProcessName);
+        foreach (Process process in processes)
+        {
+            if (process.Id != current.Id)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Uploader(string[] args)
+    {
+        traceDirectory = ParseArguments(args);
+        FileSystem fileSystem = new FileSystem();
+        config = ReadConfig(fileSystem);
+        IUpload upload = config.CreateUpload(fileSystem);
+        timerAction = new TimerAction(traceDirectory, config, upload, fileSystem);
+    }
+
+    private static Config ReadConfig(FileSystem fileSystem)
```

Comment:
I'd put all this in Config.ReadConfig. return null in case of error and exit here?

Reply:
understand...